File: 25-04-2025/4_Median_of_Two_Sorted_Arrays/solution2.js
// https://leetcode.com/problems/median-of-two-sorted-arrays/?envType=problem-list-v2&envId=array

/*
* Given two sorted arrays nums1 and nums2 of size m and n respectively,
* return the median of the two sorted arrays.
*
* The overall run time complexity should be O(log (m+n)).
*
Example 1:

Input: nums1 = [1,3], nums2 = [2]
Output: 2.00000
Explanation: merged array = [1,2,3] and median is 2.
Example 2:

Input: nums1 = [1,2], nums2 = [3,4]
Output: 2.50000
Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.

* */

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    const arr = [];
    for (let i = 0; i < nums1.length; i++) {
        arr.push(nums1[i]);
    }
    for (let i = 0; i < nums2.length; i++) {
        arr.push(nums2[i]);
    }

    for (let i = 0; i < arr.length; i++) {
        for (let j = i; j < arr.length; j++) {
            if (arr[i] > arr[j]) {
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
    }

    if (arr.length % 2 === 1) {
        return arr[1];
    } else {
        const idx = arr.length / 2;
        return (arr[idx - 1] + arr[idx]) / 2;
    }
};

console.log(findMedianSortedArrays([1, 3], [2])); // 2.00000
console.log(findMedianSortedArrays([1,2], [3,4])); // 2.50000

/*
* [1, 2, 3, 4]
* 1, 1
* 1, 2
* 1, 3
* 1, 4
*
* 2, 2
* 2, 3
* 2, 4
*
* 3, 3
* 3, 4
*
* 4, 4
* */

File: 05-05-2025/88_Merge_Sorted_Array/solution2.java
// https://leetcode.com/problems/merge-sorted-array/?envType=problem-list-v2&envId=array

// Follow up: Can you come up with an algorithm that runs in O(m + n) time?

import java.util.Arrays;

public class solution2 {
    public static void main(String[] args) {
        System.out.println(Arrays.toString(merge(new int[]{1,2,3,0,0,0}, 3, new int[]{2, 5, 6}, 3))); // {1, 2, 2, 3, 5, 6}
        System.out.println(Arrays.toString(merge(new int[]{1}, 1, new int[]{}, 0))); // {1}
        System.out.println(Arrays.toString(merge(new int[]{0}, 0, new int[]{1}, 1))); // {1}
        
        System.out.println(Arrays.toString(merge(new int[]{1,2,3,0,0}, 3, new int[]{5, 6}, 2))); // {1, 2, 2, 3, 5, 6}
    }
    
    public static int[] merge(int[] nums1, int m, int[] nums2, int n) {
        for (int j = 0; j < nums2.length; j++) {
            int idx = 0;
            for (int i = 0; i < m + j + 1 && nums2[j] > nums1[i]; i++) {
                System.out.println("nums1: " + nums1[i] + " nums2: " + nums2[j]);
                idx = i;
            }
            nums1[idx] = nums2[j];
        }
        return nums1;
    }
}
